Issue 1261 port to stampede3

wwlwpd opened the pull request:
need to test on other platforms to make sure I didn't screw something up 😁 

also, this branch contains an update to just 1 ADCIRC patchset (v53release) - once we make sure this is good on all platforms, then I will create a new issue to update the others

Testing on (checked means "OK"):

- [x] supermike III
- [x] smic
- [x] qb2
- [x] qbc
- [x] frontera
- [x] lonestar6
- [x] stampede3
- [x] ubuntu
- [x] debian
- [x] oracle linux
- [x] others??

Other things to do:

- [x] allow user to specify `serial adcirc only` (i.e., do not try to compile anything that required MPI)

wwlwpd commented on bin/init-adcirc.sh:
```diff
@@ -409,12 +409,17 @@ function dumpJSON()
     # get compiler info
-    local _FC=$(which ifort)
-    local _CC=$(which icc)
+    local _FC=$(which ifort 2> /dev/null)
```
May want to make this `ifx`

jasonfleming replied:
Most platforms are still `ifort` so this looks reasonable to me.

jasonfleming commented:
Testing on ubuntu (is that the same as debian?)

jasonfleming commented:
On the debian platform, the preprocessor `if defined` statements in the `.f90` files are not getting picked up properly. Here is the build error I got on my Ubuntu 20.04 just now:
```bash
bash init-output.sh /home/jason/Campaigns/Development/2024/issue-1261/opt gfortran
gfortran  -DASGSNETCDF -DASGSNETCDF -DHAVE_NETCDF4 -DNETCDF_CAN_DEFLATE -I. -I/home/jason/Campaigns/Development/2024/issue-1261/opt/include  -L/home/jason/Campaigns/Development/2024/issue-1261/opt/lib -c logging.f90 -L/home/jason/Campaigns/Development/2024/issue-1261/opt/lib -lnetcdff
logging.f90:621:4:

  621 | use ifcore, Only : tracebackqq
      |    1
Fatal Error: Cannot open module file ‘ifcore.mod’ for reading at (1): No such file or directory
compilation terminated.
make: *** [makefile:234: ioutil.mod] Error 1
```
it is not picking up the `if defined (INTEL)` directive.

jasonfleming commented:
Had an issue building v53release on Frontera: it is trying to use `icx` to build `adcprep`:
```bash
mpif90 -DADCNETCDF -I/work2/00976/jgflemin/frontera/issue-1261/opt/include -DHAVE_NETCDF4 -DNETCDF_CAN_DEFLATE -c -I . -I /work2/00976/jgflemin/frontera/issue-1261/opt/models/adcircs/adcirc-cg-v53release-intel-frontera/prep -O2 -FI -assume byterecl -132 -xCORE-AVX512 -DADCNETCDF -I/work2/00976/jgflemin/frontera/issue-1261/opt/include -DHAVE_NETCDF4 -DNETCDF_CAN_DEFLATE -DREAL8 -DLINUX -DCSCA -DCMPI -I /work2/00976/jgflemin/frontera/issue-1261/opt/models/adcircs/adcirc-cg-v53release-intel-frontera/work/odir4/ -o /work2/00976/jgflemin/frontera/issue-1261/opt/models/adcircs/adcirc-cg-v53release-intel-frontera/work/odir4/transport.o  /work2/00976/jgflemin/frontera/issue-1261/opt/models/adcircs/adcirc-cg-v53release-intel-frontera/src/transport.F
if [ "`echo *.mod`" != '*.mod' ]; then mv *.mod /work2/00976/jgflemin/frontera/issue-1261/opt/models/adcircs/adcirc-cg-v53release-intel-frontera/work/odir4/; fi
icx -I . -I /work2/00976/jgflemin/frontera/issue-1261/opt/models/adcircs/adcirc-cg-v53release-intel-frontera/prep -O2 -DLINUX -xCORE-AVX512 -I/work2/00976/jgflemin/frontera/issue-1261/opt/include  -c /work2/00976/jgflemin/frontera/issue-1261/opt/models/adcircs/adcirc-cg-v53release-intel-frontera/prep/mkdir.c -o /work2/00976/jgflemin/frontera/issue-1261/opt/models/adcircs/adcirc-cg-v53release-intel-frontera/work/odir4/mkdir.o
/bin/bash: icx: command not found
make[1]: *** [/work2/00976/jgflemin/frontera/issue-1261/opt/models/adcircs/adcirc-cg-v53release-intel-frontera/work/odir4/mkdir.o] Error 127
make[1]: Leaving directory `/work2/00976/jgflemin/frontera/issue-1261/opt/models/adcircs/adcirc-cg-v53release-intel-frontera/work'
make: *** [padcirc] Error 2
build failed (2)
```

wwlwpd commented:
> Had an issue building v53release on Lonestar6 and Frontera: it is trying to use `icx` to build `adcprep`:
> 
> ```shell
> mpif90 -DADCNETCDF -I/work2/00976/jgflemin/frontera/issue-1261/opt/include -DHAVE_NETCDF4 -DNETCDF_CAN_DEFLATE -c -I . -I /work2/00976/jgflemin/frontera/issue-1261/opt/models/adcircs/adcirc-cg-v53release-intel-frontera/prep -O2 -FI -assume byterecl -132 -xCORE-AVX512 -DADCNETCDF -I/work2/00976/jgflemin/frontera/issue-1261/opt/include -DHAVE_NETCDF4 -DNETCDF_CAN_DEFLATE -DREAL8 -DLINUX -DCSCA -DCMPI -I /work2/00976/jgflemin/frontera/issue-1261/opt/models/adcircs/adcirc-cg-v53release-intel-frontera/work/odir4/ -o /work2/00976/jgflemin/frontera/issue-1261/opt/models/adcircs/adcirc-cg-v53release-intel-frontera/work/odir4/transport.o  /work2/00976/jgflemin/frontera/issue-1261/opt/models/adcircs/adcirc-cg-v53release-intel-frontera/src/transport.F
> if [ "`echo *.mod`" != '*.mod' ]; then mv *.mod /work2/00976/jgflemin/frontera/issue-1261/opt/models/adcircs/adcirc-cg-v53release-intel-frontera/work/odir4/; fi
> icx -I . -I /work2/00976/jgflemin/frontera/issue-1261/opt/models/adcircs/adcirc-cg-v53release-intel-frontera/prep -O2 -DLINUX -xCORE-AVX512 -I/work2/00976/jgflemin/frontera/issue-1261/opt/include  -c /work2/00976/jgflemin/frontera/issue-1261/opt/models/adcircs/adcirc-cg-v53release-intel-frontera/prep/mkdir.c -o /work2/00976/jgflemin/frontera/issue-1261/opt/models/adcircs/adcirc-cg-v53release-intel-frontera/work/odir4/mkdir.o
> /bin/bash: icx: command not found
> make[1]: *** [/work2/00976/jgflemin/frontera/issue-1261/opt/models/adcircs/adcirc-cg-v53release-intel-frontera/work/odir4/mkdir.o] Error 127
> make[1]: Leaving directory `/work2/00976/jgflemin/frontera/issue-1261/opt/models/adcircs/adcirc-cg-v53release-intel-frontera/work'
> make: *** [padcirc] Error 2
> build failed (2)
> ```

```
ease-intel-frontera/work/odir4/ -o /work2/06482/estrabd/frontera/asgs-1261/opt/models/adcircs/adcirc-cg-v53release-intel-frontera/work/odir4/transport.o  /work2/06482/estrabd/frontera/asgs-1261/opt/models/adcircs/adcirc-cg-v53release-intel-frontera/src/transport.F
if [ "`echo *.mod`" != '*.mod' ]; then mv *.mod /work2/06482/estrabd/frontera/asgs-1261/opt/models/adcircs/adcirc-cg-v53release-intel-frontera/work/odir4/; fi
icx -I . -I /work2/06482/estrabd/frontera/asgs-1261/opt/models/adcircs/adcirc-cg-v53release-intel-frontera/prep -O2 -DLINUX -xCORE-AVX512 -I/work2/06482/estrabd/frontera/asgs-1261/opt/include  -c /work2/06482/estrabd/frontera/asgs-1261/opt/models/adcircs/adcirc-cg-v53release-intel-frontera/prep/mkdir.c -o /work2/06482/estrabd/frontera/asgs-1261/opt/models/adcircs/adcirc-cg-v53release-intel-frontera/work/odir4/mkdir.o
/bin/bash: icx: command not found
make[1]: *** [/work2/06482/estrabd/frontera/asgs-1261/opt/models/adcircs/adcirc-cg-v53release-intel-frontera/work/odir4/mkdir.o] Error 127
make[1]: Leaving directory `/work2/06482/estrabd/frontera/asgs-1261/opt/models/adcircs/adcirc-cg-v53release-intel-frontera/work'
make: *** [padcirc] Error 2
build failed (2)
```

wwlwpd commented:
@jasonfleming - fixed frontera compiling error in this latest commit, working on preprocessor for gfortran

wwlwpd commented:
@jasonfleming I think I figured it out. I had issues going on:

* I revert to c-style
* cleaned up some of the preprocessor checks
* fixed a bug that was affecting all recently added wrappers that call `make` - was not properly passing in `compiler=$ASGS_COMPILER`

I will try on ubuntu Sunday evening, but let me know if you run into any issues in the meantime.

wwlwpd commented:
saw this scroll on by, 
![image](https://github.com/StormSurgeLive/asgs/assets/46434714/beb76d0e-e2db-4753-b74d-b2f5ea9a5128)

jasonfleming commented:
> saw this scroll on by, ![image](https://private-user-images.githubusercontent.com/46434714/313447400-beb76d0e-e2db-4753-b74d-b2f5ea9a5128.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTA2NzgyMDUsIm5iZiI6MTcxMDY3NzkwNSwicGF0aCI6Ii80NjQzNDcxNC8zMTM0NDc0MDAtYmViNzZkMGUtZTJkYi00NzUzLWI3NGQtYjJmNWVhOWE1MTI4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMTclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzE3VDEyMTgyNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTZmY2FkYTc3ZTc5YWQ4YzBhNmNiYTE1MThiNmE0M2NjNDNhZGJkYzY1OWE5YjdmNzlmMWVlZDMyMGVmZmQ4ZTMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.F2Qgn93WJ7Nl6WC3uZW6n3V2nB-a2rq6750rmJ1grRo)

I'll add this to the tools issue tracker.

wwlwpd commented:
the `gfortran` option works on Frontera, should work everywhere - need to double check ADCIRC build with it on the HPCs, but if it all works there it should be fixed on ubuntu and debian

jasonfleming commented:
Yes I have confirmed that this latest branch builds on ubuntu/debian, including v53release and v55.02, and confirmed that `adcirc` and `adcprep` execute. Also confirmed that everything builds and executes on frontera and lonestar6. :-)

jasonfleming commented:
Trying this on qbc. Assuming that works, I will approve the PR unless there are other platforms I should try.

wwlwpd commented:
> Trying this on qbc. Assuming that works, I will approve the PR unless there are other platforms I should try.

I think as long as I didn't break anything else major I'll merge it and fix anything we find along the way. Ty!

wwlwpd commented:
if you can, try to build ASGS on qbc with both the default option (intel) and the `gfortran` option.

wwlwpd commented:
also I have at tweak to `init-asgs.sh` I need to commit, small and non-functional, just the elimination of a message that I am realizing is confusing

jasonfleming commented:
Was able to build and install ASGS on QBC with both intel and gfortran compilers. However, when trying to build adcirc v53release with gfortran, `make` seemed to find two sets of compiler flags that matched what it was looking for, and then somehow `ifort` ended up getting selected:
```bash
Apply patches? [yes]

applying patches from /ddnA/work/jgflemin/staging/issue-1261-gfortran/patches/ADCIRC/v53release ...
applying patch 01 ... /ddnA/work/jgflemin/staging/issue-1261-gfortran/patches/ADCIRC/v53release/01-issue-1261.patch


cd /ddnA/work/jgflemin/staging/issue-1261-gfortran/opt/models/adcircs/adcirc-cg-v53release-gfortran-queenbeeC/swan && \
   make unhcat compiler=gfortran MACHINENAME=queenbeeC DEBUG=none && \
cd /ddnA/work/jgflemin/staging/issue-1261-gfortran/opt/models/adcircs/adcirc-cg-v53release-gfortran-queenbeeC/work && \
   make padcirc adcirc adcprep hstime aswip SWAN=enable compiler=gfortran NETCDF=enable NETCDF4=enable NETCDF4_COMPRESSION=enable NETCDFHOME=/work/jgflemin/staging/issue-1261-gfortran/opt MACHINENAME=queenbeeC DEBUG=none && \
   make adcswan padcswan compiler=gfortran MACHINENAME=queenbeeC NETCDF=enable NETCDF4=enable NETCDF4_COMPRESSION=enable NETCDFHOME=/work/jgflemin/staging/issue-1261-gfortran/opt DEBUG=none

Build command contained in file, /ddnA/work/jgflemin/staging/issue-1261-gfortran/opt/models/adcircs/adcirc-cg-v53release-gfortran-queenbeeC/asgs-build.sh


JSON build info in /ddnA/work/jgflemin/staging/issue-1261-gfortran/opt/models/adcircs/adcirc-cg-v53release-gfortran-queenbeeC/adcirc.bin.buildinfo.json

/ddnA/work/jgflemin/staging/issue-1261-gfortran/opt/models/adcircs/adcirc-cg-v53release-gfortran-queenbeeC /ddnA/work/jgflemin/staging/issue-1261-gfortran/opt/models/adcircs/adcirc-cg-v53release-gfortran-queenbeeC /ddnA/work/jgflemin/staging/issue-1261-gfortran/opt/models/adcircs/adcirc-cg-v53release-gfortran-queenbeeC /ddnA/work/jgflemin/staging/issue-1261-gfortran
/ddnA/work/jgflemin/staging/issue-1261-gfortran/opt/models/adcircs/adcirc-cg-v53release-gfortran-queenbeeC /ddnA/work/jgflemin/staging/issue-1261-gfortran/opt/models/adcircs/adcirc-cg-v53release-gfortran-queenbeeC /ddnA/work/jgflemin/staging/issue-1261-gfortran

Proceed to build? [yes] 

make FOR=gfortran FFLAGS="-O -w -fno-second-underscore -ffree-line-length-none " \
        FFLAGS90="-O -w -fno-second-underscore -ffree-line-length-none -ffree-line-length-none " unhcat.exe
make[1]: Entering directory `/ddnA/work/jgflemin/staging/issue-1261-gfortran/opt/models/adcircs/adcirc-cg-v53release-gfortran-queenbeeC/swan'
gfortran HottifySWAN.f90 -c -O -w -fno-second-underscore -ffree-line-length-none -ffree-line-length-none  
gfortran HottifySWAN.o -O -w -fno-second-underscore -ffree-line-length-none  -o unhcat.exe
make[1]: Leaving directory `/ddnA/work/jgflemin/staging/issue-1261-gfortran/opt/models/adcircs/adcirc-cg-v53release-gfortran-queenbeeC/swan'
makefile:19: (INFO) Guessing the type of platform ADCIRC will run on...
makefile:31: (INFO) Name is x86_64-unknown-linux-gnu, Machine is x86_64, Vendor is unknown, and OS is linux-gnu.
makefile:32: (INFO) The root directory for the build is /ddnA/work/jgflemin/staging/issue-1261-gfortran/opt/models/adcircs/adcirc-cg-v53release-gfortran-queenbeeC
makefile:33: (INFO) The ADCIRC version is v53.05-modified
cmplrflags.mk:188: (INFO) Corresponding machine found in cmplrflags.mk.
cmplrflags.mk:421: (INFO) Corresponding machine found in cmplrflags.mk.
cmplrflags.mk:1609: (WARNING) More than one match in cmplrflags.mk. This may result in the wrong compilers being selected. Please check the cmplrflags.mk file to ensure that only one set of compiler flags is specified for your platform.
makefile:39: (INFO) The compiler variable in cmplrflags.mk is set to gfortran.
makefile:41: (INFO) The following compilers have been selected...
makefile:42: (INFO) The Fortran compiler for adcprep is set to gfortran.
makefile:43: (INFO) The serial Fortran compiler is set to gfortran.
makefile:44: (INFO) The parallel Fortran compiler is set to mpif90.
makefile:45: (INFO) The C compiler is set to gcc.
make BUILDTYPE=padcirc CC="gcc" CFLAGS="-I . -I /ddnA/work/jgflemin/staging/issue-1261-gfortran/opt/models/adcircs/adcirc-cg-v53release-gfortran-queenbeeC/prep -O2 -DLINUX" padcirc
make[1]: Entering directory `/ddnA/work/jgflemin/staging/issue-1261-gfortran/opt/models/adcircs/adcirc-cg-v53release-gfortran-queenbeeC/work'
makefile:19: (INFO) Guessing the type of platform ADCIRC will run on...
makefile:31: (INFO) Name is x86_64-unknown-linux-gnu, Machine is x86_64, Vendor is unknown, and OS is linux-gnu.
makefile:32: (INFO) The root directory for the build is /ddnA/work/jgflemin/staging/issue-1261-gfortran/opt/models/adcircs/adcirc-cg-v53release-gfortran-queenbeeC
makefile:33: (INFO) The ADCIRC version is v53.05-modified
cmplrflags.mk:188: (INFO) Corresponding machine found in cmplrflags.mk.
cmplrflags.mk:421: (INFO) Corresponding machine found in cmplrflags.mk.
cmplrflags.mk:1609: (WARNING) More than one match in cmplrflags.mk. This may result in the wrong compilers being selected. Please check the cmplrflags.mk file to ensure that only one set of compiler flags is specified for your platform.
makefile:39: (INFO) The compiler variable in cmplrflags.mk is set to gfortran.
makefile:41: (INFO) The following compilers have been selected...
makefile:42: (INFO) The Fortran compiler for adcprep is set to gfortran.
makefile:43: (INFO) The serial Fortran compiler is set to gfortran.
makefile:44: (INFO) The parallel Fortran compiler is set to mpif90.
makefile:45: (INFO) The C compiler is set to gcc.
mkdir -p /ddnA/work/jgflemin/staging/issue-1261-gfortran/opt/models/adcircs/adcirc-cg-v53release-gfortran-queenbeeC/work/odir4/
mpif90 -DADCNETCDF -I/work/jgflemin/staging/issue-1261-gfortran/opt/include -DHAVE_NETCDF4 -DNETCDF_CAN_DEFLATE -c -I . -I /ddnA/work/jgflemin/staging/issue-1261-gfortran/opt/models/adcircs/adcirc-cg-v53release-gfortran-queenbeeC/prep -O2 -ffixed-line-length-none -DADCNETCDF -I/work/jgflemin/staging/issue-1261-gfortran/opt/include -DHAVE_NETCDF4 -DNETCDF_CAN_DEFLATE -DREAL8 -DLINUX -DCSCA -DCMPI -I /ddnA/work/jgflemin/staging/issue-1261-gfortran/opt/models/adcircs/adcirc-cg-v53release-gfortran-queenbeeC/work/odir4/ -o /ddnA/work/jgflemin/staging/issue-1261-gfortran/opt/models/adcircs/adcirc-cg-v53release-gfortran-queenbeeC/work/odir4/sizes.o  /ddnA/work/jgflemin/staging/issue-1261-gfortran/opt/models/adcircs/adcirc-cg-v53release-gfortran-queenbeeC/src/sizes.F
ifort: command line warning #10006: ignoring unknown option '-ffixed-line-length-none'
/ddnA/work/jgflemin/staging/issue-1261-gfortran/opt/models/adcircs/adcirc-cg-v53release-gfortran-queenbeeC/src/sizes.F(257): error #5120: Unterminated character constant
     &                  cmdlinearg(1:4),"' is not valid and will be igno
----------------------------------------^
/ddnA/work/jgflemin/staging/issue-1261-gfortran/opt/models/adcircs/adcirc-cg-v53release-gfortran-queenbeeC/src/sizes.F(286): error #5120: Unterminated character constant
     *               cmdlinearg(1:4),"' is not valid and will be igno
-------------------------------------^
compilation aborted for /ddnA/work/jgflemin/staging/issue-1261-gfortran/opt/models/adcircs/adcirc-cg-v53release-gfortran-queenbeeC/src/sizes.F (code 1)
make[1]: *** [/ddnA/work/jgflemin/staging/issue-1261-gfortran/opt/models/adcircs/adcirc-cg-v53release-gfortran-queenbeeC/work/odir4/sizes.o] Error 1
make[1]: Leaving directory `/ddnA/work/jgflemin/staging/issue-1261-gfortran/opt/models/adcircs/adcirc-cg-v53release-gfortran-queenbeeC/work'
make: *** [padcirc] Error 2
build failed (2)
```

wwlwpd commented:
> Was able to build and install ASGS on QBC with both intel and gfortran compilers. However, when trying to build adcirc v53release with gfortran, `make` seemed to find two sets of compiler flags that matched what it was looking for, and then somehow `ifort` ended up getting selected:

Thank you, will try to reproduce it and fix what I find.

wwlwpd commented:
There is a repeated `ifdef`, removing and testing the build of just non-MPI stuff because I realize the `openmpi` setup just looks for `mpif90` in the `PATH`. It's not a real use case anyway and was pushing the boundaries, but if we wanted to make it one would take more work. Probably not what we want to do here. I'll push the commit when I am sure it's all working fine otherwise.

wwlwpd commented:
It was when `compiler` was set explicitly to `gfortran`,

```

wwlwpd commented:
testing both on frontera, then will jump over to qb2 and qbc this evening; then finally make it back round to stampede3 for a final test

wwlwpd commented:
Reconfirmed on stampede3, building with `gfortran` seems to work across the board provided you remove the `padcirc` and `padcswan` targets. The reason for this is there is no `mpif90` compatible with the system `gcc` - I suspect if we went through the trouble of setting up the environment to load the right MPI for a give `gfortran` (up to version 9), build all of ASGS with that `gfortran`, then try to build ADCIRC with _that_ `mpif90`, then it'd work fine. I am going to move this PR out of draft. Thank you for all of the help, @jasonfleming!

wwlwpd commented:
I am also testing on the oracle linux, "ASGS in a Box" from last summer's bootcamp - available here, with lab handout: https://asgs.sh/

wwlwpd commented:
@jasonfleming also let me know if you want me to run through qbc, sounded like you had it covered.

wwlwpd commented:
@jasonfleming this branch works fine on the oracle linux "ASGS in a box" VirtualBox image.

jasonfleming commented:
Ok building ASGS with intel (only) on qbc to finish off that last checkbox. Will approve this pr later this morning, once that completes successfully. :-)

jasonfleming reviewed (approved):
Works like a champ! :-)

wwlwpd commented:
Adding this to PR shortly ...

![image](https://github.com/StormSurgeLive/asgs/assets/46434714/6594a3f6-c448-4f90-be6a-7ed7d381efb0)

wwlwpd commented:
I also added info to the meta JSON that is generated via `init-adcirc.sh`

wwlwpd commented:
@jasonfleming latest commit has additional dialog in `init-adcirc.sh` to skip parallel ADCIRC (build serial) - parallel is on by default,

![image](https://github.com/StormSurgeLive/asgs/assets/46434714/0258b54f-f67f-427a-9ecc-3a7b1cef77cd)

jasonfleming reviewed (approved):
This looks great, thank you!
